# Data collection

Caio sent a file with the PDB structures mentioned in Kanteev it al, 2015: `data/PPOs-with-a-3D-structure.xlsx`.

## Seed table
A seed table was created manually: `data/seeds.tsv`.

## Fasta file
A fasta file is created by running `python src/data-collection/make-fasta.py`. This creates the file `data/seeds.fa`.

## Running Interproscan on seeds
Transfer the seeds file to the HPC: `scp data/seeds.fa idamei@transfer.gbar.dtu.dk:/work3/idamei/polyphenol-oxidases/`

On the HPC, run Interproscan:
`qrsh`
`/work3/idamei/bin/my_interproscan/interproscan-5.64-96.0/interproscan.sh -appl Pfam,SignalP_EUK,SignalP_GRAM_NEGATIVE,SignalP_GRAM_POSITIVE,Phobius -i /work3/idamei/polyphenol-oxidases/seeds.fa -f tsv -o /work3/idamei/polyphenol-oxidases/seeds.interproscan`

Download the output file:
`scp idamei@transfer.gbar.dtu.dk:/work3/idamei/polyphenol-oxidases/seeds.interproscan data/`

## Enriching seeds
The seed table is enriched with taxonomy by running `python src/data-collection/enrich-seeds.py`. This creates the file `data/seeds-enriched.tsv`.

## Trimming seeds
Align the seeds to the HMM: `hmmalign --trim data/pfam/PF00264.hmm data/seeds.fa > data/seeds.hmmalign`

Convert to fasta: `seqconverter -I stockholm -O fasta data/seeds.hmmalign > data/seeds.hmmalign.fa`.

Remove gaps `sed '/^>/!s/-//g' data/seeds.hmmalign.fa > data/seeds-trimmed.fa`

# Downloading uniprot members
Fasta of the Pfam entries was downloaded from `https://www.ebi.ac.uk/interpro/entry/pfam/PF00264/protein/UniProt/#table` and saved in `data/pfam/protein-matching-PF00264.fasta`.

A json file with metadata was also downloaded and saved in `data/pfam/protein-matching-PF00264.json`.

Aligned sequences were downloaded `data/pfam/PF00264.alignment.uniprot`.

Entries with structures were downloaded: `data/pfam/structure-matching-PF00264.tsv`.

The pfam HMM was downloaded at `https://www.ebi.ac.uk/interpro/entry/pfam/PF00264/curation/` and saved in `data/pfam/PF00264.hmm`.

A json file was downloaded with all proteome data (e-mail correspondence uniprot): `data/proteome-tree/export.json`. To download the domain data through the interproscan api, run `python src/proteome-tree/download-domain-data-api.py`. This downloads the files to `data/pfam/api`.

## Make fasta with short headers
A fasta file with only the uniprot ID is generated by running: `sed -E 's/>([A-Z0-9]+).+/>\1/' data/pfam/protein-matching-PF00264.fasta > data/pfam/protein-matching-PF00264-shortheaders.fasta`

## Make trimmed fasta with short headers
Convert alignment to fasta: `seqconverter -I stockholm -O fasta data/pfam/PF00264.alignment.uniprot > data/pfam/PF00264.alignment.uniprot.fa`.

Remove gaps `sed '/^>/!s/-//g' data/pfam/PF00264.alignment.uniprot.fa > data/pfam/PF00264-trimmed.fa`

A fasta file with only the uniprot ID as headers is generated by running: `sed -E 's/>([A-Z0-9]+).+/>\1/' data/pfam/PF00264-trimmed.fa > data/pfam/PF00264-trimmed-shortheaders.fasta`

## Enrich uniprot hits with taxonomy
To make a tsv file with taxonomy for each uniprot hit, run: `python src/data-collection/make-taxonomy-file.py`. This creates the file `data/pfam/protein-matching-PF00264.tsv`.

## Enrich uniprot hits with domain architecture
Run: `python src/proteome-tree/enrich-proteome-hits-interproscan.py`. This creates the file `data/pfam/protein-matching-PF00264-interproscan2.tsv`.

## Make protein to proteome table
Run `python src/data-collection/make-protein2proteome-table.py`. This creates the file `data/proteome-tree/protein2proteome.tsv`.

# Proteome tree

## Get proteome data
A json file with the proteomes matching PF00264 was downloaded from pfam: `data/pfam/proteome-matching-PF00264.json`.

A json file with proteome metadata was downloaded from uniprot `https://www.uniprot.org/proteomes?query=*`: `data/proteome-tree/proteomes_AND_proteome_type_1_2024_02_28.json`.

The above to files are combined by running: `python src/proteome-tree/make-proteome-table.py`. This creates the file `data/proteome-tree/proteome-data.tsv`.

## Manually selected genomes
Files with manually selected genomes were saved in `data/proteome-tree/class-representatives.xlsx` and `data/proteome-tree/fungal-order-representatives.xlsx`

## Get sequences from selected proteomes
In order to get the PPO sequences from the selected proteomes, run `python src/proteome-tree/get-proteome-sequences.py`. This creates the set of files `data/proteome-tree/selected-sequences-all-class.txt`.

## Get trimmed selected sequences
In order to make fasta files with the selected sequences, run `python src/proteome-tree/get-aligned-selected-sequences.py`. This creates the file`data/proteome-tree/all-one_proteome_per_class.trimmed.fa`.

## Make alignment
### For all kingdoms
Run linsi: `linsi --thread 7 data/proteome-tree/all-one_proteome_per_class.trimmed.fa > data/proteome-tree/all-one_proteome_per_class.trimmed-linsi.fa`

Remove gaps with more than 95% gaps: `seqconverter --remfracgapcols 0.95 -I fasta -O fasta data/proteome-tree/all-one_proteome_per_class.trimmed-linsi.fa > data/proteome-tree/all-one_proteome_per_class.trimmed-linsi-0.95.fa`

Convert to NEXUS: `seqconverter --remfracgapcols 0.95 -I fasta -O nexus data/proteome-tree/all-one_proteome_per_class.trimmed-linsi.fa > data/proteome-tree/all-one_proteome_per_class.trimmed-linsi-0.95.nexus`

### For only fungi
Run linsi: `linsi --thread 7 data/proteome-tree/fungal-one_proteome_per_order.trimmed.fa > data/proteome-tree/fungal-one_proteome_per_order.trimmed-linsi.fa`

Convert to NEXUS with removed gaps: `seqconverter --remfracgapcols 0.95 -I fasta -O nexus data/proteome-tree/fungal-one_proteome_per_order.trimmed-linsi.fa > data/proteome-tree/fungal-one_proteome_per_order.trimmed-linsi-0.95.fa`

Convert to NEXUS with removed gaps: `seqconverter --remfracgapcols 0.8 -I fasta -O nexus data/proteome-tree/fungal-one_proteome_per_order.trimmed-linsi.fa > data/proteome-tree/fungal-one_proteome_per_order.trimmed-linsi-0.8.fa`

Convert to NEXUS without removing gaps: `seqconverter -I fasta -O nexus data/proteome-tree/fungal-one_proteome_per_order.trimmed-linsi.fa > data/proteome-tree/fungal-one_proteome_per_order.trimmed-linsi.nexus`

## MrBayes
`scp data/proteome-tree/all-one_proteome_per_class.trimmed-linsi-0.95.nexus hpc:~/`
On hal, check that no jobs are running: `nvtop -d 1`

Go to the folder and start tmux: `tmux attach -d -t mrbayes`

Start the job: `mpirun -np 6 mb gpu_run.nexus > log.txt`

Detach: control+b followed by d

To continue a mrbayes run, change the number of generations to the total number of generations you want and add append=yes, fx: `mcmc append=yes ngen=60000000 samplefreq=1000 nchains=3 file=out.nex` in the run nexus file. 

The run with all kingdoms, one per order was saved in `data/mrbayes/all`.

## Clades
Files with proteins in each clade (determined from iTOL) were saved in `data/mrbayes/all/clades`.

A csv file is made by running `python src/proteome-tree/clade-annotations.py`. This creates the file `data/mrbayes/all/clades/clades.csv`.

(not used) iTOL heatmap file is made by running `python src/itol-label-files/make-itol-label-files.py` and is shown on the species tree.

A dotplot is made in R with the script `src/species-tree/dotplot.R`. 

The phylum list `data/species-tree/phylum.txt` is written in the R script `src/species-tree/dotplot.R`. 

Then, the phylum html file is written with the `src/itol-label-files/make-itol-label-files.py` script. Open the html, print the page to pdf and overlap with dotplot.

# Fingerprint (architecture figure)

## Clade data
The ids in each clade were saved in `data/mrbayes/all/clades`.

Fasta files are generated by running `python src/finger-print/make-clade-fastas.py`.

Alignments of each clade are made by running `linsi --thread 7 data/mrbayes/all-seeds-0619/clades/fastas/i_zoopago.fa > data/mrbayes/all-seeds-0619/clades/alignments/i_zoopago-linsi.fa` etc...

A fasta with all the sequences in the clades is created: `data/mrbayes/all-seeds-0619/clades/all.fa`, and an alignment is made: `linsi --thread 7 data/mrbayes/all-seeds-0619/clades/all.fa > data/mrbayes/all-seeds-0619/clades/all-linsi.fa`.

## DSSP
Pdbs are saved in `data/compare-architectures/pdbs`.

The included sequences, along with their discontinuous intervals, are listed in the script `src/compare-architectures/run-dssp.py` and this script is run in order to generate the dssp tables (saved in `data/compare-architectures/architecture-tables-dssp`).

## Aligned version
The representative sequences are retrieved from the alignment with all the data (`data/mrbayes/all-seeds-0619/clades/all-linsi.fa`) and saved in the file `data/compare-architectures/aligned-sequences.fa`.

Remove columns with all gaps: `seqconverter --remallgapcols -I fasta -O fasta data/compare-architectures/aligned-sequences.fa > data/compare-architectures/aligned-sequences-noallgaps.fa`

An architecture file is made with placeholder data that is long enough (the secondary structure will be written by cpssp, so it doesn't matter that the data is not correct, as long as it's long enough): `data/compare-architectures/aligned-new/architecture-aligned.js`.

Also a lengths file is made, the max length has to be the length of the alignment. The rest doesn't matter. `data/compare-architectures/aligned-new/lengths-aligned.js`.

## Make architecture data
Included structures are listed in the script: `src/compare-architectures/make-architecture-data.py`.

To make the architecture data, run `python src/compare-architectures/make-architecture-data.py`.

The color file is made manually: `data/compare-architectures/not-aligned-new/name2color.js`.

## CPSSP
Go to the CPSSP directory (`/Users/idamei/Downloads/cpssp.tds`)

In the CPSSP folder, a fasta file of the representatives is saved in `run-final/sequence.fa`. 

The alignment is also copied: `cp data/compare-architectures/aligned-sequences-noallgaps.fa /Users/idamei/Downloads/run-final`

Then a secondary structure fasta is made based on the `data/compare-architectures/not-aligned-new/architecture.js` file and saved in `run-final/ss.fa`.

Run `python2 tex/latex/cpssp/cpssp -s run-final/sequence.fa -u run-final/ss.fa -o run-final/out -w 35 -l 1.7`.

Upload `run-final/out0.tex` to the overleaf document. 

To make the aligned secondary structure figure, run `python2 tex/latex/cpssp/cpssp -s run-final/aligned-sequences-noallgaps.fa -u run-final/ss.fa -o run-final/aligned-out -w 35 -l 1.7`.

Upload `run-final/aligned-out0.tex` to the overleaf document.

# Lignin degraders
The list of lignin degrading species was saved in `data/lignin-degraders`.

# Pymol
To make pymol script with aligned structures and conserved residues, run `python src/structural-visualizations/make-pymol-script.py`.

To run the pymol script, run `pymol src/structural-visualizations/conserved-residues.pml`.

# iTOL label files
To make iTOL label files, run `python src/itol-label-files/make-itol-label-files.py`

# Species tree
A list of included species is made by copying it from the "class-representatives" file: `data/species-tree/species`.

This file is uploaded to timetree.org and the tree is saved in `data/species-tree/species.nwk`.

A clade file is made with underscores in species names: `data/species-tree/clades2.csv`.

The plot is made with the R script: `src/species-tree/dotplot.R`.

# Make MCC tree
`sumt --mbc -b 0.25 0.25 --biplen --rootmid -ns --basename data/epa-ng/bayes_summary -i data/mrbayes/all-seeds-0619/out.nex.run1.t -i data/mrbayes/all-seeds-0619/out.nex.run2.t`
convert to .mbc file newick in figtree and remove second tree (search for ; and delete everything thereafter). Save in `data/epa-ng/tree.nwk`

# Tree placement
Copy tree fasta: `cp data/proteome-tree/all-one_proteome_per_class.trimmed.fa data/epa-ng/ref.fa`.

Replace / with 0 to match tree file: `sed -i '' 's/\//0/g' data/epa-ng/ref.fa`.

## New seeds
Combine query and ref: `cat data/epa-ng/ref.fa data/epa-ng/new-seeds/query.fa > data/epa-ng/new-seeds/ref-query.fa`.

Make alignment: `linsi --thread 7 data/epa-ng/new-seeds/ref-query.fa > data/epa-ng/new-seeds/ref-query-linsi.fa`

Divide in two files: `data/epa-ng/new-seeds/query-linsi.fa` and `data/epa-ng/new-seeds/ref-linsi.fa`.

`mkdir data/epa-ng/new-seeds/out`

Run phylogenetic placement: `epa-ng --ref-msa data/epa-ng/new-seeds/ref-linsi.fa --tree data/epa-ng/tree.nwk --query data/epa-ng/new-seeds/query-linsi.fa --model WAG --redo --outdir data/epa-ng/new-seeds/out`

Make grafted tree: `gappa examine graft --jplace-path data/epa-ng/new-seeds/out/epa_result.jplace --fully-resolve --name-prefix gappa --out-dir data/epa-ng/new-seeds/out/ --allow-file-overwriting`. This produces the file `data/epa-ng/new-seeds/out/epa_result.newick`.

## Filtered out sequences
Copy query file: `cp data/proteome-tree/filtered-out-all-one_proteome_per_class.trimmed.fa data/epa-ng/filtered-out/query.fa`

Combine query and ref: `cat data/epa-ng/ref.fa data/epa-ng/filtered-out/query.fa > data/epa-ng/filtered-out/ref-query.fa`.

Make alignment: `linsi --thread 7 data/epa-ng/filtered-out/ref-query.fa > data/epa-ng/filtered-out/ref-query-linsi.fa`

Divide in two files: `data/epa-ng/filtered-out/query-linsi.fa` and `data/epa-ng/filtered-out/ref-linsi.fa`.

`mkdir data/epa-ng/filtered-out/out`

Run phylogenetic placement: `epa-ng --ref-msa data/epa-ng/filtered-out/ref-linsi.fa --tree data/epa-ng/tree.nwk --query data/epa-ng/filtered-out/query-linsi.fa --model WAG --redo --outdir data/epa-ng/filtered-out/out`

Make grafted tree: `gappa examine graft --jplace-path data/epa-ng/filtered-out/out/epa_result.jplace --fully-resolve --name-prefix gappa --out-dir data/epa-ng/filtered-out/out/ --allow-file-overwriting`. This produces the file `data/epa-ng/filtered-out/out/epa_result.newick`.

## Fungi one per order
Make the query file: `cat data/proteome-tree/fungi-one_proteome_per_order.trimmed.fa data/proteome-tree/filtered-out-fungi-one_proteome_per_order.trimmed.fa > data/epa-ng/fungi-order/query.fa`
Replace / with 0 to match tree file: `sed -i '' 's/\//0/g' data/epa-ng/fungi-order/query.fa`.

Combine query and ref: `cat data/epa-ng/ref.fa data/epa-ng/fungi-order/query.fa > data/epa-ng/fungi-order/ref-query.fa`.

Make alignment: `linsi --thread 7 data/epa-ng/fungi-order/ref-query.fa > data/epa-ng/fungi-order/ref-query-linsi.fa`

Divide in two files: `data/epa-ng/fungi-order/query-linsi.fa` and `data/epa-ng/fungi-order/ref-linsi.fa`.

Run phylogenetic placement: `epa-ng --ref-msa data/epa-ng/fungi-order/ref-linsi.fa --tree data/epa-ng/tree.nwk --query data/epa-ng/fungi-order/query-linsi.fa --model WAG --redo --outdir data/epa-ng/fungi-order/out`

Make grafted tree: `gappa examine graft --jplace-path data/epa-ng/fungi-order/out/epa_result.jplace --fully-resolve --name-prefix gappa --out-dir data/epa-ng/fungi-order/out/`. This produces the file `data/epa-ng/fungi-order/out/epa_result.newick`.

## Place seeds on short fungal tree
Make MCC tree: `sumt --mbc -b 0.25 0.25 --biplen --rootmid -ns --basename data/epa-ng/seeds-short-fungal/bayes_summary -i data/mrbayes/short-fungal-0507/out.nex.run1.t -i data/mrbayes/short-fungal-0507/out.nex.run2.t`
convert .mbc file newick in figtree and remove second tree (search for ; and delete everything thereafter). Save in `data/epa-ng/short-fungal-0507/tree.nwk`

(old) Copy tree fasta: `cp data/proteome-tree/fungal-one_proteome_per_order.trimmed.fa data/epa-ng/seeds-short-fungal/ref.fa`.
Replace / with 0 to match tree file: `sed -i '' 's/\//0/g' data/epa-ng/seeds-short-fungal/ref.fa`.

(new) Make tree fasta: `seqconverter --degap -I nexus -O fasta data/mrbayes/short-fungal-0507/short-fungal-linsi-95.nexus > data/epa-ng/seeds-short-fungal/ref.fa`

Make query file with short fungal seeds: `data/epa-ng/seeds-short-fungal/query.fa`

Combine query and ref: `cat data/epa-ng/seeds-short-fungal/ref.fa data/epa-ng/seeds-short-fungal/query.fa > data/epa-ng/seeds-short-fungal/ref-query.fa`.

Make alignment: `linsi --thread 7 data/epa-ng/seeds-short-fungal/ref-query.fa > data/epa-ng/seeds-short-fungal/ref-query-linsi.fa`

Divide in two files: `data/epa-ng/seeds-short-fungal/query-linsi.fa` and `data/epa-ng/seeds-short-fungal/ref-linsi.fa`.

Run phylogenetic placement: `epa-ng --ref-msa data/epa-ng/seeds-short-fungal/ref-linsi.fa --tree data/epa-ng/seeds-short-fungal/tree.nwk --query data/epa-ng/seeds-short-fungal/query-linsi.fa --model WAG --redo --outdir data/epa-ng/seeds-short-fungal/out`

Make grafted tree: `gappa examine graft --jplace-path data/epa-ng/seeds-short-fungal/out/epa_result.jplace --fully-resolve --name-prefix gappa --out-dir data/epa-ng/seeds-short-fungal/out/`. This produces the file `data/epa-ng/seeds-short-fungal/out/epa_result.newick`.

# Tree of short fungal
The accs in the short fungal clade are retrieved from the tree and saved in `data/short-fungal-tree/accs`.

A fasta is made by running: `python src/short-fungal-tree/get-sequences.py`

Alignment: `linsi --thread 7 data/short-fungal-tree/short-fungal.fasta > data/short-fungal-tree/short-fungal-linsi.fasta`

Convert to nexus: `seqconverter -I fasta -O nexus data/short-fungal-tree/short-fungal-linsi.fasta > data/short-fungal-tree/short-fungal-linsi.nexus`

The Mrbayes tree is run on hal and saved in `data/mrbayes/short-fungal-0507`.

# Boxplots of number of PPOs per class
To make the data table for making boxplots, run `python src/boxplots/make-proteome-table-with-empty.py`. This creates the file `data/boxplots/boxplot-data.tsv`.

To make the boxplots, run the R script `src/boxplots/make-boxplots.R`.

