# Data collection

Caio sent a file with the PDB structures mentioned in Kanteev it al, 2015: `data/PPOs-with-a-3D-structure.xlsx`.

## Seed table
A seed table was created manually: `data/seeds.tsv`.

## Fasta file
A fasta file is created by running `python src/data-collection/make-fasta.py`. This creates the file `data/seeds.fa`.

## Running Interproscan on seeds
Transfer the seeds file to the HPC: `scp data/seeds.fa idamei@transfer.gbar.dtu.dk:/work3/idamei/polyphenol-oxidases/`

On the HPC, run Interproscan:
`/work3/idamei/bin/my_interproscan/interproscan-5.64-96.0/interproscan.sh -appl Pfam,SignalP_EUK,SignalP_GRAM_NEGATIVE,SignalP_GRAM_POSITIVE,Phobius -i /work3/idamei/polyphenol-oxidases/seeds.fa -f tsv -o /work3/idamei/polyphenol-oxidases/seeds.interproscan`

Download the output file:
`scp idamei@transfer.gbar.dtu.dk:/work3/idamei/polyphenol-oxidases/seeds.interproscan data/`

## Enriching seeds
The seed table is enriched with taxonomy by running `python src/data-collection/enrich-seeds.py`. This creates the file `data/seeds-enriched.tsv`.

## Trimming seeds
Align the seeds to the HMM: `hmmalign --trim data/pfam/PF00264.hmm data/seeds.fa > data/seeds.hmmalign`

Convert to fasta: `seqconverter -I stockholm -O fasta data/seeds.hmmalign > data/seeds.hmmalign.fa`.

Remove gaps `sed '/^>/!s/-//g' data/seeds.hmmalign.fa > data/seeds-trimmed.fa`

# Downloading uniprot members
Fasta of the Pfam entries was downloaded from `https://www.ebi.ac.uk/interpro/entry/pfam/PF00264/protein/UniProt/#table` and saved in `data/pfam/protein-matching-PF00264.fasta`.

A json file with metadata was also downloaded and saved in `data/pfam/protein-matching-PF00264.json`.

(Not used) Aligned sequences were downloaded `data/pfam/PF00264.alignment.uniprot`.

Entries with structures were downloaded: `data/pfam/structure-matching-PF00264.tsv`.

The pfam HMM was downloaded at `https://www.ebi.ac.uk/interpro/entry/pfam/PF00264/curation/` and saved in `data/pfam/PF00264.hmm`.

A json file was downloaded with all proteome data (e-mail correspondence uniprot): `data/proteome-tree/export.json`. To download the domain data through the interproscan api, run `python src/proteome-tree/download-domain-data-api.py`. This downloads the files to `data/pfam/api`.

## Make fasta with short headers
A fasta file with only the uniprot ID is generated by running: `sed -E 's/>([A-Z0-9]+).+/>\1/' data/pfam/protein-matching-PF00264.fasta > data/pfam/protein-matching-PF00264-shortheaders.fasta`

## Make trimmed fasta with short headers
Convert alignment to fasta: `seqconverter -I stockholm -O fasta data/pfam/PF00264.alignment.uniprot > data/pfam/PF00264.alignment.uniprot.fa`.

Remove gaps `sed '/^>/!s/-//g' data/pfam/PF00264.alignment.uniprot.fa > data/pfam/PF00264-trimmed.fa`

A fasta file with only the uniprot ID as headers is generated by running: `sed -E 's/>([A-Z0-9]+).+/>\1/' data/pfam/PF00264-trimmed.fa > data/pfam/PF00264-trimmed-shortheaders.fasta`

## Enrich uniprot hits with taxonomy
To make a tsv file with taxonomy for each uniprot hit, run: `python src/data-collection/make-taxonomy-file.py`. This creates the file `data/pfam/protein-matching-PF00264.tsv`.

## Enrich uniprot hits with domain architecture
Run: `python src/proteome-tree/enrich-proteome-hits-interproscan.py`. This creates the file `data/pfam/protein-matching-PF00264-interproscan2.tsv`.

# Proteome tree

## Get proteome data
A json file with the proteomes matching PF00264 was downloaded from pfam: `data/pfam/proteome-matching-PF00264.json`.

A json file with proteome metadata was downloaded from uniprot `https://www.uniprot.org/proteomes?query=*`: `data/proteome-tree/proteomes_AND_proteome_type_1_2024_02_28.json`.

The above to files are combined by running: `python src/proteome-tree/make-proteome-table.py`. This creates the file `data/proteome-tree/proteome-data.tsv`.

## Manually selected genomes
Files with manually selected genomes were saved in `data/proteome-tree/class-representatives.xlsx` and `data/proteome-tree/fungal-order-representatives.xlsx`

## Get sequences from selected proteomes
In order to get the PPO sequences from the selected proteomes, run `python src/proteome-tree/get-proteome-sequences.py`. This creates the set of files `data/proteome-tree/selected-proteomes-ids-fungi-order.txt` and `data/proteome-tree/fungal-one_proteome_per_order.fa`.

## Get trimmed selected sequences
In order to make fasta files with the selected sequences, run `python src/proteome-tree/get-aligned-selected-sequences.py`. This creates the files `data/proteome-tree/selected-sequences-all-class.txt` and `data/proteome-tree/selected-proteomes-ids-all-class.txt`.

## Make alignment
### For all kingdoms
Run linsi: `linsi --thread 7 data/proteome-tree/all-one_proteome_per_class.trimmed.fa > data/proteome-tree/all-one_proteome_per_class.trimmed-linsi.fa`

Remove gaps with more than 95% gaps: `seqconverter --remfracgapcols 0.95 -I fasta -O fasta data/proteome-tree/all-one_proteome_per_class.trimmed-linsi.fa > data/proteome-tree/all-one_proteome_per_class.trimmed-linsi-0.95.fa`

Convert to NEXUS: `seqconverter --remfracgapcols 0.95 -I fasta -O nexus data/proteome-tree/all-one_proteome_per_class.trimmed-linsi.fa > data/proteome-tree/all-one_proteome_per_class.trimmed-linsi-0.95.fa`

### For only fungi
Run linsi: `linsi --thread 7 data/proteome-tree/fungal-one_proteome_per_order.trimmed.fa > data/proteome-tree/fungal-one_proteome_per_order.trimmed-linsi.fa`

Convert to NEXUS with removed gaps: `seqconverter --remfracgapcols 0.95 -I fasta -O nexus data/proteome-tree/fungal-one_proteome_per_order.trimmed-linsi.fa > data/proteome-tree/fungal-one_proteome_per_order.trimmed-linsi-0.95.fa`

Convert to NEXUS with removed gaps: `seqconverter --remfracgapcols 0.8 -I fasta -O nexus data/proteome-tree/fungal-one_proteome_per_order.trimmed-linsi.fa > data/proteome-tree/fungal-one_proteome_per_order.trimmed-linsi-0.8.fa`

Convert to NEXUS without removing gaps: `seqconverter -I fasta -O nexus data/proteome-tree/fungal-one_proteome_per_order.trimmed-linsi.fa > data/proteome-tree/fungal-one_proteome_per_order.trimmed-linsi.nexus`

## MrBayes
On hal, go to the folder and start tmux: `tmux attach -d -t mrbayes`

Start the job: `mpirun -np 6 mb gpu_run.nexus > log.txt`

Detach: control+b followed by d

To continue a mrbayes run, change the number of generations to the total number of generations you want and add append=yes, fx: `mcmc append=yes ngen=60000000 samplefreq=1000 nchains=3 file=out.nex` in the run nexus file. 
(old, don't do this) copy everything from the .ckp~ file into the bottom of the aignment nexus file.

The run with all kingdoms, one per order was saved in `data/mrbayes/all`.

## Clades
Files with proteins in each clade (determined from iTOL) were saved in `data/mrbayes/all/clades`.

A csv file is made by running `python src/proteome-tree/clade-annotations.py`. This creates the file `data/mrbayes/all/clades/clades.csv`.

iTOL heatmap file is made by running `python src/itol-label-files/make-itol-label-files.py` and is shown on the species tree.

# Fingerprint
The ids in each clade were saved in `data/mrbayes/all/clades`.

Fasta files are generated by running `python src/finger-print/make-clade-fastas.py`.

# Lignin degraders
The list of lignin degrading species was saved in `data/lignin-degraders`.

# Pymol
To make pymol script with aligned structures and conserved residues, run `python src/structural-visualizations/make-pymol-script.py`.

To run the pymol script, run `pymol src/structural-visualizations/conserved-residues.pml`.

# iTOL label files
To make iTOL label files, run `python src/itol-label-files/make-itol-label-files.py`

# Species tree
A list of included species is made: `data/species-tree/species`.

This file is uploaded to timetree.org and the tree is saved in `data/species-tree/species.nwk`.

A clade file is made with underscores in species names: `data/species-tree/clades2.csv`.

The plot is made with the R script: `src/species-tree/dotplot.R`.